### PR TITLE
Fixes job handling involving functions and terminal control

### LINKED ISSUE #3805: When called from another shell script, fish can't read
- [x] Have you checked if problem occurs with [fish 2.4.0](/fish-shell/fish-shell/releases/tag/2.4.0)?
- [x] Tried fish without third-party customizations *(check `sh -c 'env HOME=$(mktemp -d) fish'`)*?

**fish version installed** *(`fish --version`)*: fish, version 2.3.1

**OS/terminal used**: Debian Linux (jessie), xterm

I wrote a script using `#!/usr/bin/fish`.  The script calls `read`.  Works fine from command line but not when called from another script.

## Reproduction steps

````
nr@homedog ~/net> call-fish-read 
fish: Couldn't grab control of terminal
tcsetpgrp: Interrupted system call

nr@homedog ~/net> fish-read
say something⏎                                                                  asdf

asdf
nr@homedog ~/net> /usr/bin/fish --version
fish, version 2.3.1-dirty
````

Scripts are attached in a zip file, but it may be easier to view source here.

Here is `fish-read`:

````
#!/usr/bin/fish

read -p 'echo -n say something >&2 ' -l -s answer
echo $answer
````

and here is `call-fish-read`:

````
#!/bin/ksh

mumble=$(fish-read)
echo $mumble
````

N.B. I downloaded a nightly of fish 2.4.0 from SUSE, and behavior is even more frustrating---the shell just hangs trying to read file descriptor 3.  Not sure what's up with that (I poked at it with `strace -f`.)

Would love to have a workaround that enables me to use a fish script as a tool in other scripts.
[demo-scripts.zip](https://github.com/fish-shell/fish-shell/files/749392/demo-scripts.zip) 

May be related to #3223 and #1258.

### LINKED ISSUE #3952: Job handling is broken.
I have an issue running the command `sudo echo lol | grep lol`. This command is not as trivial as it seems because grep is a fish function, calling `command grep --color=auto $argv` behind the scene.

Now, a simple run of that command gives
```
layus@klatch ~/p/fish> sudo echo lol | grep lol
# Nothing happens, job is blocked. Entering ^Z.
Job 1, 'sudo echo lol | grep lol' has stopped
layus@klatch ~/p/fish> jobs
Job	Group	CPU	State	Command
1	7318	0%	stopped	sudo echo lol | grep lol
2	7319	0%	stopped	command grep --color=auto $argv
layus@klatch ~/p/fish> fg
Send job 1, 'sudo echo lol | grep lol' to foreground
[sudo] password for layus : 
layus@klatch ~/p/fish> fg
Send job 2, 'command grep --color=auto $argv' to foreground
lol
layus@klatch ~/p/fish> 
```

This behavior is not changed by #3922, and may relate to #206. We see that fish starts two jobs for one user job, and that these jobs are managed separately. Digging in the source code, we can find that one job is created for the top-level pipeline.

[exec_job](https://github.com/fish-shell/fish-shell/blob/c0de8afaf34471478db83b38f9994e6d937e6517/src/exec.cpp#L368-L1130) starts processes for each step of the pipeline. When it reaches `grep`, it detects a fish function and calls [internal_exec_helper](https://github.com/fish-shell/fish-shell/blob/c0de8afaf34471478db83b38f9994e6d937e6517/src/exec.cpp#L667-L668) to start its body.
That function parses the function body and starts a new job to to handle its content.
That new job is not aware of its parent context, and start with a new pgid. Fish gives that process control over the terminal, removing it from `sudo`.

When `sudo` starts, it receives a SIGTTOU and gets stopped. Fish blocks because the grep job is still running, and unrelated to the sudo job.

Running ps to compare fish to other shells gives

<pre>
 PPID   PID  PGID   SID TTY      TPGID STAT   UID   TIME COMMAND
===== ===== ===== ===== ======== ===== ===== ==== ====== =========
    1   764   764   764 ?           -1 Ss       0   0:00 sshd [...]
  764 31954 31954 31954 ?           -1 Ss       0   0:00  \_ sshd: me [priv]
31954 31956 31954 31954 ?           -1 S     1000   0:00  |   \_ sshd: me@pts/0
31956 31957 31957 31957 pts/0     7319 Ss    1000   0:00  |       \_ -zsh
31957  7306  7306 31957 pts/0     7319 S     1000   0:00  |           \_ ./fish -d4
 7306  7317  7306 31957 pts/0     7319 S     1000   0:00  |               \_ ./fish -d4
 7306  7318  <b>7318</b> 31957 pts/0     <b>7319 T</b>        0   0:00  |               \_ sudo echo lol
 7306  7319  <b>7319</b> 31957 pts/0     <b>7319</b> S+    1000   0:00  |               \_ grep --color=auto lol
  764  7390  7390  7390 ?           -1 Ss       0   0:00  \_ sshd: me [priv]
 7390  7392  7390  7390 ?           -1 S     1000   0:00  |   \_ sshd: me@pts/4
 7392  7393  7393  7393 pts/4     7470 Ss    1000   0:00  |       \_ -zsh
 7393  7460  7460  7393 pts/4     7470 S     1000   0:00  |           \_ zsh
 7460  7470  <b>7470</b>  7393 pts/4     <b>7470</b> S+       0   0:00  |               \_ sudo echo lol
 7460  7471  <b>7470</b>  7393 pts/4     <b>7470</b> S+    1000   0:00  |               \_ grep --color=auto lol
  764  7474  7474  7474 ?           -1 Ss       0   0:00  \_ sshd: me [priv]
 7474  7476  7474  7474 ?           -1 S     1000   0:00      \_ sshd: me@pts/5
 7476  7477  7477  7477 pts/5     7530 Ss    1000   0:00          \_ -zsh
 7477  7519  7519  7477 pts/5     7530 S     1000   0:00              \_ bash
 7519  7530  <b>7530</b>  7477 pts/5      <b>7530</b> S+       0   0:00                  \_ sudo echo lol
 7519  7531  <b>7530</b>  7477 pts/5      <b>7530</b> S+    1000   0:00                  \_ grep lol
</pre>

We can see that the current TPGID is 7319 while the sudo command has a PGID of 7318. It's status is T, which means that the process is "stopped, either by a job control signal or because it is being traced". Here it is because of the aforementioned SIGTTOU.
We see that other shells give the same PGID to all the subcommands.

The extra fish process is the "keepalive fork" used by fish. I still wonder why fish needs it while other shells do not, but it is not important.

From a job control point of view, fish shows two running jobs in `jobs` output, while bash and zsh will show only one, even if zsh seems to distinguish between the processes of the job. (^Z was added for clarity, but does not appear in the terminal)

<pre>
ZSH:
--(layus@klatch)--> grep () { command grep --color=auto "$@"; }   
--(layus@klatch)-->  sudo echo lol | grep lol                      
[sudo] password for layus: <em>^Z</em>
zsh: running    sudo echo lol | grep --color=auto lol
--(layus@klatch)--> jobs
[1]  + suspended           sudo echo lol | 
       suspended (signal)  grep --color=auto lol

BASH:
layus@klatch:~$ grep () { command grep --color=auto "$@"; }
layus@klatch:~$ sudo echo lol | grep lol
[sudo] password for layus: <em>^Z</em>

[1]+  Stopped                 sudo echo lol | grep lol
layus@klatch:~$ jobs
[1]+  Stopped                 sudo echo lol | grep lol
</pre>

We need to make function invocation part of the same job as their context. I believe that one command line should create one job in the jobs view. That would also immediately fix the SIGTTOU issue for sudo because both processes will have the same PGID and will both have control over the terminal.
Because they are handled exctly like function, begin/end blocks would be fixed at the same time.

---

On current master (version 2.5.0-259-gc0de8afa-dirty, aka c0de8afa).

### PR DESCRIPTION
**Pull request by mqudsi**
While the idiomatic fix to fish' myriad of job control issues would be
to parse all functions prior to beginning the job pipeline so that
everything in the command line can be executed in the context of a
single job, that would require a huge effort to rewrite the core job
flow in fish and does not make sense at this time.

Instead, this patch fixes #3952 and #206 (but notably not #4238) by
having jobs that are part of a single command pipeline, including those
that are functions executing external commands, use the same process
group. This prevents a (parent|child) from crashing with SIGTTIN or
hanging at SIGTTOU because it has a different process group than the
process currently in control of the terminal.

Additionally, since this fix involves removing the code that forces fish
to run in its own process group (which IMHO never made sense, as job
control is the job of the shell, not the process being launched), it
also fixes #3805 and works around BashOnWindows#1653.

### CODE REVIEW

**Comment by faho** [comment]
This does _not_ fix #206, since the issue there is that we lose the redirection.

In

```fish
function testfun
    set input (cat)
    echo $input
end

echo testing123 | testfun
```

the `cat` should be connected to the _pipe_ (and hence read "testing123"), not the terminal.

**Comment by faho** [comment]
It does seem to fix #3952.

Fixing #4238 would require adding the parent job to the newly created one.

I have actually had a slightly different idea - store the process group of a job that has "JOB_TERMINAL" and just use it for all jobs launched while that is active. That should fix both of them, IIUC.

------

As a sidenote, we prefer the imperative mood in commit messages - e.g. this would be "Fix job handling...". It's a bit shorter and keeps the focus on the commit ("if applied, this commit  will").

**Comment by mqudsi** [comment]
The original #206 report is hard to follow. This addresses at least one is the bugs in the comments which I did not find reported elsewhere. On a phone right no2w, I'll update again soon.

**Comment by faho** [comment]
Ah, sorry!

#206 is hard to follow because it was half-fixed in #3922. Previously, the `(cat)` would actually result in a stopped cat, because fish never did the pgroup dance for it to begin with.

That then caused people to jump on and report things that were better handled in different issues. The `cat | meh` thing is basically #3952, the `read n` thing is #3805.

**Comment by krader1961** [comment]
I don't see anything obviously wrong with the change. I'll play with it a bit just to confirm I don't see any new issues and that it does fix the stated issues.

File: src/reader.cpp [hunk 1]
```diff
@@ -1626,30 +1626,7 @@ static void reader_interactive_init() {
...
-        if (errno == ENOTTY) redirect_tty_output();
-        debug(0, _(L"Couldn't grab control of terminal"));
-        wperror(L"tcsetpgrp");
-        exit_without_destructors(1);
-    }
-
     invalidate_termsize();
-
-    // Set the new modes.
-    if (tcsetattr(0, TCSANOW, &shell_modes) == -1) {
```
Comment (zanchey) [inline src/reader.cpp]: For my understanding, why shouldn't fish set terminal modes here? I know it already happens in `reader_readline` - is that enough?

Comment (mqudsi) [inline src/reader.cpp]: Unfortunately both `tcsetattr` and `tcsetpgrp` are responsible for deadlocking fish/the reader since it seems that just like fish sometimes spawns fish and waits for the process to be in a certain state before giving it control of the shell, other shells (`bash -c`) do too.

The man page for `tcgetattr` specifies it can be used from background processes, but the docs on `tcsetattr` are very skimpy and don't mention anything at all. It _seems_ from our experience that this isn't safe to call until the shell has full control of the terminal.

In the interest of laying out all options, it might be possible to use SIGALRM to force an interrupt in both these calls if they haven't responded in xx seconds, but I would rather just call them once we know we have control of the terminal at some later point.

**Comment by mqudsi** [comment]
I really want to fix all the "definitely broken" core job handling cases; imho if a shell can't get job handling right, nothing else matters.

@faho I think you're right that keeping the `JOB_TERMINAL` pgrp for all subsequent jobs while that one still lives would fix the problem, but I fear it may also encompass jobs that aren't part of the same pipeline? The real problem is, of course, in the parser that breaks a pipeline down into jobs well before it reaches `exec_job`, anything we do in `exec_job` is just a workaround for broken job tokenization (and that's not code I've looked at, though I suspect it's a giant spaghetti state machine that I would prefer not to modify if at all possible).

**Comment by faho** [comment]
>@faho I think you're right that keeping the JOB_TERMINAL pgrp for all subsequent jobs while that one still lives would fix the problem, but I fear it may also encompass jobs that aren't part of the same pipeline?

The question here is what job _should_ get the terminal. And currently, every job that is not part of an event gets "JOB_TERMINAL" (see `parse_execution_context_t::run_1_job`), so they _should_ get it. Personally, I'd prefer just using that and tightening it later if it turns out to be too lenient. Maybe also use "JOB_FOREGROUND".

I think having a job in the foreground pgroup that shouldn't be is less of a PITA than the other way around - which is what we currently have.

>The real problem is, of course, in the parser that breaks a pipeline down into jobs well before it reaches exec_job, anything we do in exec_job is just a workaround for broken job tokenization

Yes.

**Comment by mqudsi** [comment]
Ugh this is a very ugly situation. My previous overhaul that added signal synchronization (`SIGSTOP` and `SIGCONT` marshaling in `exec_job`) to job control was to eliminate the cases where fish just hopes it can randomly decide to join a process group that may or may not still be around by the time the `tcsetpgrp` call was made. There is no such synchronization across groups, the assumption being that a group is a logical grouping of all interdependent processes. 

Trying to join the pgrp of a previous process from a different job means needing to handle not just the cases when the process is still running or already exited, but also dealing correctly with the cases where the process exits after you check if it's running but before the call to `tcsetpgrp` to join its process group goes through.

This patch only handled cases where there was recursion into `exec_job`, (e.g. a job that contains a function/builtin call which starts a new fish instance to evaluate it), but it seems the needed solution would have to extend the lifetime of the cached process group well past the end of `exec_job`'s lifetime for the first job in the pipeline.

I had a straightforward solution that involved replacing the call to `terminal_return_from_job` with a check that tests for any other foreground processes that may still be running (indicating another job that is part of the same pipeline, since this was only called when foreground processes were exiting) and, if they exist, first transferring control of the terminal to _that_ process group via `terminal_give_to_job` or `job_continue` (which calls `terminal_give_to_job`) but the problem is that fish's checks to see if there are any child processes running are really fragile (or just plain incorrect) and didn't play well with this change,  leading to either "failed to regain control of terminal" (`ENOTTY`) errors or deadlocks with 100% CPU usage as fish goes into a `waitpid` loop with incomplete termination logic.

**Comment by layus** [comment]
As this is an improvement over the current situation, can we still consider merging it? I went in the details of the existing implementation and it's indeed gigantic and messy. I would prefer a quick fix for now, because it seems we will not find anyone to rewrite that code any time soon.

Also, seeing patches waiting validation forever is a strong deterrent for contributors ;)

**Comment by faho** [comment]
>Also, seeing patches waiting validation forever is a strong deterrent for contributors ;)

Note that @mqudsi has commit privileges. He can merge it whenever he pleases.

We're also currently preparing a release (2.7.0) that won't have this in it because it's too risky. So there's no pressure to merge.

Though, yes, from what I've seen this improves the situation and should be merged unless it makes further improvement harder.

**Comment by layus** [comment]
> Note that @mqudsi has commit privileges. He can merge it whenever he pleases.
> 
> We're also currently preparing a release (2.7.0) that won't have this in it because it's too risky. So there's no pressure to merge.

Ah, I did not know that. Thanks for clarifying.

**Comment by mqudsi** [comment]
I was holding off because I was trying to find a cleaner way instead of adding what amounts to a hack that only works around part of the problem.. but that hasn't gone very well.

I've been testing this locally on a few machines on different operating systems and I suspect there may be a bug in it (shell hangs if background child process in control of terminal is interrupted with a signal or some other mess like that), but I need to double-check whether it's a regression or not. I really don't want to merge anything that will make matters worse and the fact that it's not a rewrite (where I know exactly what I'm doing and can precisely track down the effect of changes) means we could be introducing subtle (bad) changes if we're not careful.

**Comment by mqudsi** [comment]
Yeah, there's a new bug introduced by this that's not in `master`. I can reproduce it locally by launching fzf history search via its ctrl+r keybinding then using ctrl+c to abort. That hangs the shell in a loop (cpu usage, not deadlock) somewhere.

**Comment by mqudsi** [comment]
I just merged 55b3c45f95ba09efa4b9eaad8ec49256a7435823 which includes a subset of this patch to close #3805 and fix the hang on startup issue under unknown situations under WSL.

I'll continue to explore a fix for the hangs discussed earlier in this thread, but an easy fix is looking less and less likely.
